src/theory/quantifiers/sygus/sygus_reconstruct.cpp

```diff
@@ -0,0 +1,533 @@
+  // satisfied
+  unsolvedObs[stn].emplace(mainOb);
+  d_obInfo.emplace(mainOb, RConsObligationInfo(sol));
+  d_stnInfo[stn].setBuiltinToOb(sol, mainOb);
+
+  // We need to add the main obligation to the crd in case it cannot be broken
+  // down by matching. By doing so, we can solve the obligation using
+  // enumeration and crd (if it is in the grammar)
+  d_stnInfo[stn].addTerm(sol);
+
+  // the set of unique (up to rewriting) patterns/shapes in the grammar used by
+  // matching
+  std::unordered_map<TypeNode, std::vector<Node>, TypeNodeHashFunction> pool;
+
+  size_t count = 0;
+
+  // algorithm
+  while (d_sol[mainOb] == Node::null() && count < enumLimit)
+  {
+    // enumeration phase
+    // a temporary set of new obligations cached for processing in the match
+    // phase
+    std::unordered_map<TypeNode,
+                       std::unordered_set<Node, NodeHashFunction>,
+                       TypeNodeHashFunction>
+        obsPrime;
+    for (const std::pair<const TypeNode,
+                         std::unordered_set<Node, NodeHashFunction>>& pair :
+         unsolvedObs)
+    {
+      // enumerate a new term
+      Trace("sygus-rcons") << "enum: " << stn << ": ";
+      Node sz = d_stnInfo[pair.first].nextEnum();
+      if (sz == Node::null())
+      {
+        continue;
+      }
+      Node builtin = Rewriter::rewrite(datatypes::utils::sygusToBuiltin(sz));
+      // if enumerated term does not contain free variables, then its
+      // corresponding obligation can be solved immediately
+      if (sz.isConst())
+      {
+        Node rep = d_stnInfo[pair.first].addTerm(builtin);
+        Node k = d_stnInfo[pair.first].builtinToOb(rep);
+        // check if the enumerated term solves an obligation
+        if (k == Node::null())
+        {
+          // if not, create an "artifical" obligation whose solution would be
+          // the enumerated term
+          k = nm->mkSkolem("sygus_rcons", pair.first);
+          d_obInfo.emplace(k, RConsObligationInfo(builtin));
+          d_stnInfo[pair.first].setBuiltinToOb(builtin, k);
+        }
+        // mark the obligation as solved
+        markSolved(k, sz);
+        // with the candidate rewrite database, there is no point in adding
+        // ground terms to the pool
+        continue;
+      }
+      // if there are no matches (all calls to notify return true)...
+      if (d_poolTrie.getMatches(builtin, this))
+      {
+        // then, this is a new term and we should add it to pool
+        d_poolTrie.addTerm(builtin);
+        pool[pair.first].push_back(sz);
+        for (Node k : pair.second)
+        {
+          if (d_sol[k] == Node::null())
+          {
+            Trace("sygus-rcons")
+                << "ob: " << RConsObligationInfo::obToString(k, d_obInfo[k])
+                << std::endl;
+            // try to match obligation k with the enumerated term sz
+            std::unordered_map<TypeNode,
```

Comment:
Can you introduce a typedef for this?  This would be a using in the header of the class, for and example see https://github.com/CVC4/CVC4/blob/master/src/printer/let_binding.h#L76